obsolete/powerd.8
'\" -*- coding: UTF-8 -*-
.\" Copyright (C) 1994 Miquel van Smoorenburg.
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
.\"
.TH POWERD 8 "Feb 14, 1994" "" "Linux System Administrator's Manual"
.SH NAME
.\" powerd \(em monitor a serial line connected to an UPS.
powerd -- monitor a serial line connected to an UPS.
.SH SYNOPSIS
.B /sbin/powerd
.RB " serial-device "
.SH DESCRIPTION
.B Powerd
is a daemon process that sits in the background and monitors the state
of the DCD line of the serial device. This line is meant to be
connected to a UPS (Uninterruptible Power Supply) so that \fBpowerd\fP knows
about the state of the UPS. As soon as \fBpowerd\fP senses that the
power is failing (it sees that DCD goes low) it notifies \fBinit\fP(8),
and \fBinit\fP then executes the \fBpowerwait\fP and \fBpowerfail\fP entries.
If \fBpowerd\fP senses that the power has been restored, it notifies \fBinit\fP
again and \fBinit\fP will execute the \fBpowerokwait\fP entries.
.SH ARGUMENTS
.IP serial-device
Some serial port that is not being used by some other device, and does not
share an interrupt with any other serial port.
.SH DIAGNOSTICS
\fBPowerd\fP regularly checks the \fBDSR\fP line to see if it's high.
\fBDSR\fP should be directly connected to \fBDTR\fP and \fBpowerd\fP
keeps that line high, so if \fBDSR\fP is low then something is wrong
with the connection. \fBPowerd\fP will notify you about this fact every
two minutes. When it sees that the connection has been restored it
will say so.
.SH HOWTO
It's pretty simple to connect your UPS to the Linux machine. The steps
are easy:
.TP 0.5i
.B 1.
Make sure you have an UPS with a simple relay output: it should
close its connections (make) if the power is gone, and it should
open its connections (break) if the power is good.
.TP 0.5i
.B 2.
Buy a serial plug. Connect the DTR line to the DSR line directly.
Connect the DTR line and the DCD line with a \fB10 kilo ohm\fP
resistor. Now connect the relay output of the UPS to GROUND
and the DCD line. If you don't know what pins DSR, DTR, DCD and
GROUND are you can always ask at the store where you bought the plug.
.TP 0.5i
.B 3.
You're all set.
.SH BUGS
Well, not a real bug but \fBpowerd\fP should be able to do a broadcast or
something on the ethernet in case more Linux-boxes are connected to
the same UPS and only one of them is connected to the UPS status line.
.SH SEE ALSO
.BR shutdown (8),
.BR init (8),
.BR inittab (5)
.SH AUTHOR
Miquel van Smoorenburg, miquels@cistron.nl


doc/Changelog
sysvinit (2.87dsf) world; urgency=low

  * Fix typos and do minor updates in the manual pages.
  * Correct section of mountpoint(1).
  * Document -e and -t options for telinit in init(8).
  * Update address of FSF in the COPYRIGHT file.
  * Document in halt(8) that -n might not disable all syncing.
    Patch by Bill Nottingham and Fedora
  * Adjust output from "last -x".  In reboot lines, print endpoint
    of uptime too.  In shutdown lines print downtimes rather than
    the time between downs.  Fix typo in string compare in last.c.
    Patch by Thomas Hood.
  * Improve handling of IPv6 addresses in last.  Patch from Fedora.
  * Document last options in usage information, previously only
    mentioned in the manual page.
  * Add new option -F to last, to output full date string instead
    of the short form provided by default.  Patch from Olaf Dabrunz
    and SuSe.
  * Adjust build rules to make sure the installed binaries
    are stripped.
  * Increase the compiler warning level when building.
  * Fix utmp/wtmp updating on 64-bit platforms.  Patch by Bill
    Nottingham and Fedora.
  * Avoid unchecked return value from malloc() in utmpdump.
    Patch from Christian 'Dr. Disk' Hechelmann and Fedora.
  * Make sure to use execle and no execl when passing environment to
    the new process.  Patch from RedHat.
  * Correct init to make sure the waiting status is preserved across
    re-exec.  Patch from RedHat.
  * Correct init to avoid race condition when starting programs during
    boot.  Patch from SuSe.
  * Allow 'telinit u' in runlevels 0 and 6.  Patch from Thomas Hood.
  * Change install rules to make pidof an absolute symlink.  Patch from
    Thomas Hood.
  * Improve error message from init if fork() fail.  Patch found in Suse.
  * Add support for SE Linux capability handling.  Patch from Manoj
    Srivastava, adjusted to avoid aborting if SE policy was loaded in
    the initrd with patch from Bill Nottingham and Fedora.
  * Add -c option to pidof for only matching processes with the same
    process root.  Ignore -c when not running as root.  Patch from
    Thomas Woerner and Fedora.
  * Adjust init to terminate argv0 with one 0 rather than two so that
    process name can be one character longer.  Patch by Kir Kolyshkin.
  * Make sure bootlogd exit with non-error exit code when forking of
    the child successfully.
  * Add bootlogd option -s to make it possible to control the use of
    fdatasync().  Patch from Thomas Hood.
  * Add bootlogd option -c to tell it to create the log file if it does
    not exist.  Patch from Thomas Hood.
  * Let bootlogd also look at ttyB* devices to work on HPPA.  Patch
    from Thomas Hood.
  * Change init to use setenv() instead of putenv, make sure the PATH
    value is usable on re-exec.  Patch from Thomas Hood.
  * Add usleep in killall5 after killing processes, to force the kernel
    to reschedule.  Patch from SuSe.
  * Modify pidof to not print empty line if no pid was found.
  * Modify init and sulogin to fix emergency mode's tty, making sure ^C
    and ^Z work when booting with 'emergency' kernel option.  Patch from
    Samuel Thibault.
  * Modify init to allow some time for failed opens to resolve themselves.
    Patch from Bill Nottingham and Fedora.
  * Modify init to shut down IDE, SCSI and SATA disks properly.  Patches
    from Sebastian Reichelt, Werner Fink and SuSe.
  * Modify wall to use UT_LINESIZE from <utmp.h> instead of hardcoded
    string lengths.  Patch from SuSe.
  * Change wall to make halt include hostname in output.
  * Change killall to avoid killing init by mistake.  Patch from SuSe.
  * Change killall5 to use the exit value to report if it found any
    processes to kill.  Patch from Debian.
  * Add option -o opmitpid to killall5, to make it possible to skip
    some pids during shutdown.  Based on patch from Colin Watson and
    Ubuntu.
  * Add references between killall5 and pidof manual pages.  Patch from Debian.
  * Modify killall to work better with user space file system, by
    changing cwd to /proc when stopping and killing processes, and
    avoiding stat() when the value isn't used.  Also, lock process
    pages in memory to avoid paging when user processes are stopped.
    Patch from Debian and Goswin von Brederlow with changes by Kel
    Modderman.
  * Change shutdown to only accept flags -H and -P with the -h flag,
    and document this requirement in the manual page.
  * Change reboot/halt to work properly when used as a login shell.
    Patch by Dale R. Worley and Fedora.
  * Let sulogin fall back to the staticly linked /bin/sash if both roots
    shell and /bin/sh fail to execute.

 -- Petter Reinholdtsen <pere@hungry.com>  Sun, 12 Jul 2009 19:58:10 +0200

sysvinit (2.86) cistron; urgency=low

  * Fixed up bootlogd to read /proc/cmdline. Also keep an internal
    linebuffer to process \r, \t and ^H. It is becoming useable.
  * Applied trivial OWL patches
  * Block signals in syslog(), since syslog() is not re-entrant
    (James Olin Oden <joden@malachi.lee.k12.nc.us>, redhat bug #97534)
  * Minor adjustements so that sysvinit compiles on the Hurd 
  * killall5 now skips kernel threads
  * Inittab entries with both 'S' and other runlevels were broken.
    Fix by Bryan Kadzban <bryan@kadzban.is-a-geek.net>
  * Changed initreq.h to be more flexible and forwards-compatible.
  * You can now through /dev/initctl set environment variables in
    init that will be inherited by its children. For now, only
    variables prefixed with INIT_ can be set and the maximum is
    16 variables. There's also a length limit due to the size
    of struct init_request, so it should be safe from abuse.
  * Option -P and -H to shutdown set INIT_HALT=POWERDOWN and
    INIT_HALT=HALT as environment variables as described above
  * Add "mountpoint" utility.
  * Slightly better algorithm in killall5.c:pidof()
  * Added some patches from fedora-core (halt-usage, last -t,
    sulogin-message, user-console)


obsolete/powerd.c
/*
 * powerd	Monitor the DCD line of a serial port connected to
 *		an UPS. If the power goes down, notify init.
 *		If the power comes up again, notify init again.
 *		As long as the power is OK, the DCD line should be
 *		"HIGH". When the power fails, DCD should go "LOW".
 *		Powerd keeps DTR high so that you can connect
 *		DCD and DTR with a resistor of 10 Kilo Ohm and let the
 *		UPS or some relais pull the DCD line to ground.
 *		You also need to connect DTR and DSR together. This
 *		way, powerd can check now and then if DSR is high
 *		so it knows the UPS is connected!!
 *
 * Usage:	powerd /dev/cua4 (or any other serial device).
 *
 * Author:	Miquel van Smoorenburg, <miquels@drinkel.cistron.nl>.
 *
 * Version:	1.31,  29-Feb-1996.
 *
 *		This program was originally written for my employer,
 *			** Cistron Electronics **
 *		who has given kind permission to release this program
 *		for general puppose.
 *
 *		Copyright (C) 1991-1996 Cistron Electronics.
 *
 *		This program is free software; you can redistribute it and/or modify
 *		it under the terms of the GNU General Public License as published by
 *		the Free Software Foundation; either version 2 of the License, or
 *		(at your option) any later version.
 *
 *		This program is distributed in the hope that it will be useful,
 *		but WITHOUT ANY WARRANTY; without even the implied warranty of
 *		MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *		GNU General Public License for more details.
 *
 *		You should have received a copy of the GNU General Public License
 *		along with this program; if not, write to the Free Software
 *		Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

/* Use the new way of communicating with init. */
#define NEWINIT

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <fcntl.h>
#include <errno.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <signal.h>
#include <syslog.h>
#include <string.h>
#include "paths.h"
#ifdef NEWINIT
#include "initreq.h"
#endif

#ifndef SIGPWR
#  define SIGPWR SIGUSR1
#endif

#ifdef NEWINIT
void alrm_handler()
{
}
#endif

/* Tell init the power has either gone or is back. */
void powerfail(ok)
int ok;
{
  int fd;
#ifdef NEWINIT
  struct init_request req;

  /* Fill out the request struct. */
  memset(&req, 0, sizeof(req));
  req.magic = INIT_MAGIC;
  req.cmd   = ok ? INIT_CMD_POWEROK : INIT_CMD_POWERFAIL;

  /* Open the fifo (with timeout) */
  signal(SIGALRM, alrm_handler);
  alarm(3);
  if ((fd = open(INIT_FIFO, O_WRONLY)) >= 0
		&& write(fd, &req, sizeof(req)) == sizeof(req)) {
	close(fd);
	return;
  }
  /* Fall through to the old method.. */
#endif

  /* Create an info file for init. */
  unlink(PWRSTAT);
  if ((fd = open(PWRSTAT, O_CREAT|O_WRONLY, 0644)) >= 0) {
	if (ok)
		write(fd, "OK\n", 3);
	else
		write(fd, "FAIL\n", 5);
	close(fd);
  }
  kill(1, SIGPWR);
}

/* Main program. */
int main(int argc, char **argv)
{
  int fd;
  int dtr_bit = TIOCM_DTR;
  int flags;
  int status, oldstat = -1;
  int count = 0;
  int tries = 0;

  if (argc < 2) {
	fprintf(stderr, "Usage: powerd <device>\n");
	exit(1);
  }

  /* Start syslog. */
  openlog("powerd", LOG_CONS|LOG_PERROR, LOG_DAEMON);

  /* Open monitor device. */
  if ((fd = open(argv[1], O_RDWR | O_NDELAY)) < 0) {
	syslog(LOG_ERR, "%s: %s", argv[1], sys_errlist[errno]);
	closelog();
	exit(1);
  }

  /* Line is opened, so DTR is high. Force it anyway to be sure. */
  ioctl(fd, TIOCMBIS, &dtr_bit);

  /* Daemonize. */
  switch(fork()) {
	case 0: /* Child */
		closelog();
		setsid();
		break;
	case -1: /* Error */
		syslog(LOG_ERR, "can't fork.");
		closelog();
		exit(1);
	default: /* Parent */
		closelog();
		exit(0);
  }

  /* Restart syslog. */
  openlog("powerd", LOG_CONS, LOG_DAEMON);

  /* Now sample the DCD line. */
  while(1) {
	/* Get the status. */
	ioctl(fd, TIOCMGET, &flags);

	/* Check the connection: DSR should be high. */
	tries = 0;
	while((flags & TIOCM_DSR) == 0) {
		/* Keep on trying, and warn every two minutes. */
		if ((tries % 60) == 0)
		    syslog(LOG_ALERT, "UPS connection error");
		sleep(2);
		tries++;
  		ioctl(fd, TIOCMGET, &flags);
	}
	if (tries > 0)
		syslog(LOG_ALERT, "UPS connection OK");

	/* Calculate present status. */
	status = (flags & TIOCM_CAR);

	/* Did DCD drop to zero? Then the power has failed. */
	if (oldstat != 0 && status == 0) {
		count++;
		if (count > 3)
			powerfail(0);
		else {
			sleep(1);
			continue;
		}
	}
	/* Did DCD come up again? Then the power is back. */
	if (oldstat == 0 && status > 0) {
		count++;
		if (count > 3)
			powerfail(1);
		else {
			sleep(1);
			continue;
		}
	}
	/* Reset count, remember status and sleep 2 seconds. */
	count = 0;
	oldstat = status;
	sleep(2);
  }
  /* Never happens */
  return(0);
}
